### PR TITLE
spec: Add spec items for environment fallback hosts

### LINKED ISSUE #818: Client libraries to generate fallback hosts from custom environment
This enhancement will automatically use fallback hosts, deterministically derived from the supplied `environment`, when they've not been specified explicitly.

It will solve the problem whereby customers ship their app built against our client library, seeing that it works with basic testing, but don't realise that it will quickly stop working when things go wrong with the primary host.

See the [internal issue](https://github.com/ably/wiki/issues/361) which proposes this enhancement.

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/DOC-58) by [Unito](https://www.unito.io/learn-more)

### PR DESCRIPTION
**Pull request by lmars**
This adds spec items describing how clients should generate fallback hosts when `ClientOptions#environment` is set (which take the form `<environment>-[a-e]-fallback.ably-realtime.com`).

It also deprecates the `ClientOptions#fallbackHostsUseDefault` because it is unlikely that if `ClientOptions#restHost` or `ClientOptions#environment` are explicitly set that the default fallbacks should be used.

Closes #818.

### CODE REVIEW

File: content/client-lib-development-guide/features.textile [hunk 1]
```diff
@@ -91,13 +91,22 @@ h3(#restclient). RestClient
...
 * @(RSC13)@ The client library must use the connection and request timeouts specified in the @ClientOptions@, falling back to the defaults described in @ClientOptions@ below
 * @(RSC15)@ Host Fallback
-** @(RSC15b)@ The fallback behavior described below only applies when the default @rest.ably.io@ endpoint is being used and has not been overriden (see "RSC11":#RSC11), @ClientOptions#fallbackHostsUseDefault@ is @true@, or an array of @ClientOptions#fallbackHosts@ is provided. If host fallback is not supported, failing HTTP requests that would have "qualified for a retry against a fallback host (see RSC15d)":#RSC15d, will instead result in an error immediately
+** @(RSC15b)@ The fallback behavior described below only applies when either @ClientOptions#restHost@ has not been set to an explicit value (see "RSC11a":#RSC11a), the deprecated @ClientOptions#fallbackHostsUseDefault@ option is set to @true@, or an array of @ClientOptions#fallbackHosts@ is provided. If host fallback is not supported, failing HTTP requests that would have "qualified for a retry against a fallback host (see RSC15d)":#RSC15d, will instead result in an error immediately
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile:100]: "is not supported" -> "does not apply" ?

File: content/client-lib-development-guide/features.textile [hunk 2]
```diff
@@ -91,13 +91,22 @@ h3(#restclient). RestClient
...
+** @(RSC15a)@ In the case of an error necessitating use of an alternative host (see "RSC15d":#RSC15d), try fallback hosts (with a matching Host header as this is necessary when fallbacks are proxied through a CDN) in random order, continuing to try further hosts if "qualifying errors":#RSC15d occur, failing when all have been tried or the configured @httpMaxRetryCount@ has been reached (see "TO3l@":#TO3l5). This ensures that a client library is able to work around routing or other problems for the user's closest datacenter. For example, if a @POST@ request to @rest.ably.io@ fails because the default endpoint is unreachable or unserviceable, then the @POST@ request should be retried again against the fallback hosts in attempt to find an alternate healthy datacenter to service the request
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile:102]: I know that this text was already there, but do we really need to say that you include a host header for a given target? Or if it does need to be said, maybe it is stated as a standalone requirement somewhere?

File: content/client-lib-development-guide/features.textile [hunk 3]
```diff
@@ -91,13 +91,22 @@ h3(#restclient). RestClient
...
+** @(RSC15g)@ Fallback hosts are chosen as follows:
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile:103]: Maybe ftaod this should state explicitly "When the use of fallbacks applies, the set of fallback hosts is chosen ...".

File: content/client-lib-development-guide/features.textile [hunk 4]
```diff
@@ -91,13 +91,22 @@ h3(#restclient). RestClient
...
+*** @(RSC15g1)@ If @ClientOptions#fallbackHosts@ is set to an array, use those as the fallback hosts. See "TO3k6":#TO3k6 for constraints
+*** @(RSC15g2)@ If @ClientOptions#environment@ is set to a value other than "production" and @ClientOptions#fallbackHosts@ is not set, use the environment fallback hosts (see "RSC15i":#RSC15i)
+*** @(RSC15g3)@ If @ClientOptions#fallbackHosts@ and @ClientOptions#environment@ are not set, use the default fallback hosts (see "RSC15h":#RSC15h)
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile:106]: If both ...

File: content/client-lib-development-guide/features.textile [hunk 5]
```diff
@@ -454,9 +463,9 @@ h3(#realtime-connection). Connection
...
+** @(RTN17a)@ By default, every connection attempt is first attempted to the default primary host @realtime.ably.io@ (unless overriden in @ClientOptions#environment@ or @ClientOptions#realtimeHost@), which, through DNS, is automatically routed to the client's closest datacenter. The client library must always prefer the default endpoint (closest datacenter), even if a previous connection attempt to that endpoint has failed. (That is, @RSC15f@ does not apply)
+** @(RTN17c)@ In the case of an error necessitating use of an alternative host (see "RTN17d":#RTN17d), the @Connection@ manager should first check if an internet connection is available by issuing a @GET@ request to @https://internet-up.ably-realtime.com/is-the-internet-up.txt@. If the request succeeds and the text "yes" is included in the body, then the client library can assume it has a viable internet connection and should then immediately retry the connection against fallback hosts (with a matching Host header as this is necessary when fallbacks are proxied through a CDN) in random order to find an alternative healthy datacenter. Fallback hosts are chosen as per "RSC15g":#RSC15g
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile:468]: Same comment as above re Host header

**Review by paddybyers — COMMENTED** [review]
A few small comments

File: content/client-lib-development-guide/features.textile (hunk 1 above)
Comment (lmars) [reply content/client-lib-development-guide/features.textile]: Updated in 88d5ace1

File: content/client-lib-development-guide/features.textile (hunk 2 above)
Comment (lmars) [reply content/client-lib-development-guide/features.textile]: I'm going off the assumption that it was added for a reason, and I don't think it does any harm do be explicit, so I've moved it out into `RSC15j` in 482fba5b.

File: content/client-lib-development-guide/features.textile (hunk 3 above)
Comment (lmars) [reply content/client-lib-development-guide/features.textile]: Yes I think that's better, updated in 24ddec36

File: content/client-lib-development-guide/features.textile (hunk 4 above)
Comment (lmars) [reply content/client-lib-development-guide/features.textile]: Added in 24ddec36

File: content/client-lib-development-guide/features.textile (hunk 5 above)
Comment (lmars) [reply content/client-lib-development-guide/features.textile]: Addressed in 482fba5b

**Review by paddybyers — APPROVED** [review]
LGTM, thanks

**Review by SimonWoolf — APPROVED** [review]

**Pull request merged**